### PR TITLE
Spanify work, 1 of x

### PR DESCRIPTION
**Pull request by iamcarbon**
The PR is focused on spanifying the first round of classes and methods in preparation to replace IList<byte> and reduce allocations more broadly through the library. 

The big changes here are introducing polyfills for GetString(ReadOnlySpan<byte>) on older frameworks, introducing a ArrayPoolBufferWriter (to utilize IBufferWriter), and spanifying various methods that will be called in future PRs.

### CODE REVIEW

**Comment by iamcarbon** [comment]
@BobLd Ready for initial feedback. This PR sets up the next round of optimizations.

File: src/UglyToad.PdfPig.Tokens/HexToken.cs [hunk 1]
```diff
@@ -106,7 +113,7 @@ public HexToken(IReadOnlyList<char> characters)
         /// <param name="high">The high nibble.</param>
         /// <param name="low">The low nibble.</param>
         /// <returns>The byte.</returns>
-        public static byte Convert(char high, char low)
+        public static byte ConvertPair(char high, char low)
```
Comment (iamcarbon) [inline src/UglyToad.PdfPig.Tokens/HexToken.cs:116]: This was renamed to avoid a conflict with Convert.ToHexString

**Comment by iamcarbon** [comment]
"Passed!  - Failed:     0, Passed:  3256, Skipped:     6, Total:  3262, Duration: 1 m 48 s - UglyToad.PdfPig.Tests.dll (net8.0)"

Tests seem to be passing, but build is still failing. Will look into this more later.

File: src/UglyToad.PdfPig.Core/Polyfills/EncodingExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿#if NETFRAMEWORK || NETSTANDARD2_0
+
+namespace System.Text;
+
+internal static class EncodingExtensions
+{
+    public unsafe static string GetString(this Encoding encoding, ReadOnlySpan<byte> bytes)
+    {
```
Comment (BobLd) [inline src/UglyToad.PdfPig.Core/Polyfills/EncodingExtensions.cs:8]: if (bytes.Length == 0)
        {
            return string.Empty;
        }

File: src/UglyToad.PdfPig.Tokens/HexToken.cs [hunk 3]
```diff
@@ -10,8 +10,7 @@ namespace UglyToad.PdfPig.Tokens
     /// </summary>
     public class HexToken : IDataToken<string>
```
Comment (BobLd) [inline src/UglyToad.PdfPig.Tokens/HexToken.cs:11]: can be sealed?

**Comment by BobLd** [comment]
@iamcarbon thanks for the great job, looks good to me so far. The only area where I'm not sure is about using unsafe code - let avoid that for the moment. If not using unsafe become a real performance bottleneck for net framework, let's do that in a later PR

**Comment by BobLd** [comment]
@iamcarbon are you willing to merge this PR? I see you flagged it as [Draft]

**Comment by iamcarbon** [comment]
@BobLd Yep, this is good to go from my side. Ready to merge when everyone is happy.

**Review by BobLd — APPROVED** [review]

**Pull request merged**